cluster-autoscaler/hack/update-mods.sh

```diff
@@ -0,0 +1,206 @@
+
+    ## Add staging module overrides.
+    ## Extraneous replacements are harmless
+    for mod in ${K8S_STAGING_MODULES[*]}; do
+        go mod edit -require=${mod}@v0.0.0
+        go mod edit -replace=${mod}=k8s.io/kubernetes/staging/src/${mod}@${K8S_PSEUDO_VERSION}
+    done
+fi
+
+################################################################################
+### Get k8s.io/kubernetes/go.mod@$K8S_REV ######################################
+################################################################################
+cd $WORK_DIR
+K8S_GO_MOD_JSON="k8s-go-mod.json"
+curl -O $WORK_DIR/go.mod -s "https://raw.githubusercontent.com/${K8S_FORK}/${K8S_REV}/go.mod"
+go mod edit -json > $WORK_DIR/$K8S_GO_MOD_JSON
+
+################################################################################
+### Ensure k8s.io/kubernetes version and overrides in mod file #################
+################################################################################
+
+# Get Cluster Autoscaler required packages
+cd $AUTOSCALE_DIR
+CA_GO_MOD_JSON="ca-go-mod.json"
+go mod edit -json > $WORK_DIR/$CA_GO_MOD_JSON
+CA_REQS=$(cat $WORK_DIR/$CA_GO_MOD_JSON | jq -r .Require[].Path)
+
+# Record Cluster Autoscaler differences from k8s/k8s
+CA_DIFFS=()
+CA_ONLY_MODS=()
+echo "Cluster Autoscale & Kubernetes Dependencies"
+for req in $CA_REQS; do
+    # Get corresponding dependency version from k8s/k8s
+    K8S_DEPENDENCY_VERSION=$(cat $WORK_DIR/$K8S_GO_MOD_JSON | jq --arg req $req -r '.Require[] | select(.Path == $req) | .Version')
+
+    # Skip modules staged in k8s/k8s
+    SKIP="0"
+    for STAGED_MOD in ${K8S_STAGING_MODULES[*]}; do
+        if [ "$req" == "$STAGED_MOD" ]; then
+            SKIP="1"
+        fi
+    done
+    # Skip packages that are permitted to drift
+    for ALLOWED_DIFF in ${PERMITTED_PACKAGE_DIFFS[*]}; do
+        if [ "$req" == "$ALLOWED_DIFF" ]; then
+            SKIP="1"
+        fi
+    done
+
+    CA_VERSION=$(cat $WORK_DIR/$CA_GO_MOD_JSON | jq -r --arg req $req '.Require[] | select(.Path == $req) | .Version')
+    # if cluster-autoscaler requires it and k8s/k8s does, use k8s/k8s version
+    if [ -n "$K8S_DEPENDENCY_VERSION"  ] && [ "$SKIP" -eq "0"  ]; then
+        if [ "$CA_VERSION" != "$K8S_DEPENDENCY_VERSION" ]; then
+            CA_DIFFS+=("${req}@${CA_VERSION} != ${req}@${K8S_DEPENDENCY_VERSION}")
+        else
+            echo "    ${req}@${K8S_DEPENDENCY_VERSION}"
+        fi
+
+        if [ "$UPDATE_GO_MOD" == "true" ]; then
+            echo "    * Pinning kubernetes dependency in autoscaler to ${req}@${K8S_DEPENDENCY_VERSION}"
+            # pin dependency to k8s/k8s & force replace
+            go mod edit -require=${req}@${K8S_DEPENDENCY_VERSION}
+            go mod edit -replace=${req}=${req}@${K8S_DEPENDENCY_VERSION}
+        fi
+    fi
+    if [ -z "$K8S_DEPENDENCY_VERSION" ]; then
+        CA_ONLY_MODS+=("${req}@${CA_VERSION}")
+    fi
+done
+
+# TODO Exit if CA_DIFFS is not empty?
```

Comment:
What would happen if go.mod file contains a dependency `A` in version `V1` as a transitive dependency of some k8s code and on call to `update-mods.sh` the version of `A` required by k8s is bumped to `V2`.
It feels like we will have `A` in `CA_DIFFS` then. 

That is normal and expected situation. We should not warn user about that. 

The disturbing situation is when we are using `Replace` for `A` in `go.mod`, to use version which is different than one used in `k8s`. Or more specifically older than one used in `k8s`. IMO we should error out only in such case.

Reply:
> What would happen if go.mod file contains a dependency A in version V1 as a transitive dependency of some k8s code and on call to update-mods.sh the version of A required by k8s is bumped to V2.
It feels like we will have A in CA_DIFFS then.

It will, but just below with `UPDATE_GO_MOD="true"`, we'll always set the `require` and the `replace` to whatever k8s uses

```bash
go mod edit -require=${req}@${K8S_DEPENDENCY_VERSION} 2&1       
go mod edit -replace=${req}=${req}@${K8S_DEPENDENCY_VERSION} 2&1
```

> That is normal and expected situation. We should not warn user about that.

I changed this to only report any detected drift from k8s/k8s when go.mod is not edited

> The disturbing situation is when we are using Replace for A in go.mod, to use version which is different than one used in k8s. Or more specifically older than one used in k8s. IMO we should error out only in such case.

This shouldn't happen with the above 

```bash
go mod edit -require=${req}@${K8S_DEPENDENCY_VERSION} 2&1       
go mod edit -replace=${req}=${req}@${K8S_DEPENDENCY_VERSION} 2&1
```